airflow/models.py

```diff
@@ -1447,12 +1474,20 @@ def signal_handler(signum, frame):
                 Stats.incr('operator_successes_{}'.format(
                     self.task.__class__.__name__), 1, 1)
                 Stats.incr('ti_successes')
+            self.refresh_from_db(lock_for_update=True)
             self.state = State.SUCCESS
         except AirflowSkipException:
+            self.refresh_from_db(lock_for_update=True)
             self.state = State.SKIPPED
         except (Exception, KeyboardInterrupt) as e:
-            self.handle_failure(e, test_mode, context)
-            raise
+            self.refresh_from_db()
+            # for case when task is marked as success externally
+            # current behavior doesn't hit the success callback
+            if self.state == State.SUCCESS:
```

Comment:
The behavior being indistiguishable becaise in the case where it is success, the monitoring process would have shut it down anyway, possibly before this exception is thrown.